content/en/blog/2025/ai-agent-observability/index.md

```diff
@@ -0,0 +1,225 @@
+[emit traces, metrics, and logs](/docs/concepts/instrumentation/).
+
+Different AI Agent frameworks have varying approaches to implementing
+observability, mainly categorized into two options:
+
+#### Option 1: Baked-in Instrumentation
+
+The first option is to implement built-in instrumentation that emits telemetry
+using OpenTelemetry semantic conventions. This means observability is a native
+feature, allowing users to seamlessly track agent performance, task execution,
+and resource utilization. Some AI Agent frameworks, such as CrewAI, follow this
+pattern.
+
+As a developer of an agent framework, here are some pros and cons of this
+baked-in instrumentation:
+
+- Pros
+  - You can take on the maintenance overhead of keeping the instrumentation for
+    telemetry up-to-date.
+  - Simplifies adoption for users unfamiliar with OpenTelemetry configuration.
+- Cons
+  - Adds bloat to the framework for users who do not need observability
+    features.
+  - Risk of version lock-in if the framework’s OpenTelemetry dependencies lag
+    behind upstream updates.
+  - Less flexibility for advanced users who prefer custom instrumentation.
+- Some best practices to follow if you consider this approach:
+  - Provide a configuration setting that lets users easily enable or disable
+    telemetry collection from your framework's built-in instrumentation.
+  - Plan ahead of users wanting to use other external instrumentation packages
+    and avoid collision.
+  - Consider listing your agent framework in the
+    [OpenTelemetry registry](/ecosystem/registry/) if you choose this path.
+- As a developer of an agent application, you may want to choose an agent
+  framework with baked-in instrumentation if you prefer…
+  - Minimal dependencies on external packages in your agent app code
+  - Out-of-the-box observability without manual setup.
+
+#### Option 2: Instrumentation via OpenTelemetry contrib
+
+The second option is to publish language-specific libraries to the OpenTelemetry
+contrib libraries (ex:
+[opentelemetry-instrumentation-vertexai](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-vertexai)).
+These libraries can be imported into agents and configured to emit telemetry per
+OpenTelemetry semantic conventions.
+
+As a developer of an agent framework, here are some pros and cons of this
+baked-in instrumentation:
+
+- Pros
+  - Decouples observability from the core framework, reducing bloat.
+  - Leverages OpenTelemetry’s community-driven maintenance for instrumentation
+    updates.
+  - Allows users to mix and match contrib libraries for their specific needs
+    (e.g., cloud providers, LLM vendors).
+- Cons
+  - Users must manually install and configure contrib libraries, increasing
+    setup complexity.
+  - Risk of fragmentation if users rely on incompatible or outdated contrib
+    packages.
+  - Less control over telemetry quality and coverage compared to baked-in
```

Comment:
True, though:

 1. The quality is likely to be higher given the higher bar held in that repo for telemetry quality.
 2. There is still the ability to contribute to it in order to improve it.

I think the "less control" is more relevant in relation to velocity rather than in relation to quality.

Reply:
good point, updated to below

```
Development velocity slows down when there are too many PRs in the OpenTelemetry review queue.
```